File: code/cli/dimensions/src/dimensions/formatter.py
"""
Output formatting functionality for displaying dimension analysis results.
"""

import sys
from dataclasses import asdict
from pathlib import Path
from typing import Dict, List, Optional
import structlog

import orjson
import yaml
from rich.console import Console
from rich.table import Table
from rich.text import Text

from .processor import DimensionStats
from .utils import format_file_size, truncate_file_list


class OutputFormatter:
    """Formats and outputs dimension analysis results."""
    
    def __init__(self, logger: structlog.BoundLogger):
        """
        Initialize the output formatter.
        
        Args:
            logger: Structured logger instance
        """
        self.logger = logger
        self.console = Console(file=sys.stderr)  # Rich output goes to stderr
    
    def format_results(self,
                      results: Dict[str, DimensionStats],
                      format_type: str = "text",
                      sort_by: str = "count",
                      min_count: int = 1,
                      max_results: Optional[int] = None,
                      output_file: Optional[Path] = None) -> None:
        """
        Format and output results in the specified format.
        
        Args:
            results: Dictionary of dimension statistics
            format_type: Output format ('text', 'json', 'yaml')
            sort_by: Sort criteria ('count' or 'dimensions')
            min_count: Minimum count to include
            max_results: Maximum number of results to show
            output_file: Optional file to save output
        """
        # Filter and sort results
        filtered_results = self._filter_and_sort_results(
            results, sort_by, min_count, max_results
        )
        
        # Generate summary statistics
        summary = self._generate_summary(results, filtered_results)
        
        # Format based on type
        if format_type == "text":
            self._output_text_format(filtered_results, summary, output_file)
        elif format_type == "json":
            self._output_json_format(filtered_results, summary, output_file)
        elif format_type == "yaml":
            self._output_yaml_format(filtered_results, summary, output_file)
        else:
            raise ValueError(f"Unsupported format type: {format_type}")
    
    def _filter_and_sort_results(self,
                                results: Dict[str, DimensionStats],
                                sort_by: str,
                                min_count: int,
                                max_results: Optional[int]) -> List[DimensionStats]:
        """
        Filter and sort dimension statistics.
        
        Args:
            results: Raw results dictionary
            sort_by: Sort criteria ('count' or 'dimensions')
            min_count: Minimum count filter
            max_results: Maximum results limit
            
        Returns:
            Filtered and sorted list of DimensionStats
        """
        # Filter by minimum count
        filtered = [stats for stats in results.values() if stats.count >= min_count]
        
        # Sort based on criteria
        if sort_by == "count":
            filtered.sort(key=lambda x: x.count, reverse=True)
        elif sort_by == "dimensions":
            filtered.sort(key=lambda x: (x.width, x.height))
        else:
            raise ValueError(f"Unsupported sort criteria: {sort_by}")
        
        # Limit results
        if max_results is not None:
            filtered = filtered[:max_results]
        
        return filtered
    
    def _generate_summary(self,
                         all_results: Dict[str, DimensionStats],
                         filtered_results: List[DimensionStats]) -> Dict:
        """
        Generate summary statistics.
        
        Args:
            all_results: All dimension statistics
            filtered_results: Filtered results being displayed
            
        Returns:
            Summary statistics dictionary
        """
        total_images = sum(stats.count for stats in all_results.values())
        total_size = sum(stats.total_size for stats in all_results.values())
        unique_dimensions = len(all_results)
        
        # Find most common dimension
        most_common = None
        if all_results:
            most_common = max(all_results.values(), key=lambda x: x.count)
        
        return {
            "total_images": total_images,
            "unique_dimensions": unique_dimensions,
            "total_size_bytes": total_size,
            "total_size": format_file_size(total_size),
            "most_common_dimension": most_common.dimensions_str if most_common else "None",
            "most_common_count": most_common.count if most_common else 0,
            "displayed_results": len(filtered_results)
        }
    
    def _output_text_format(self,
                           results: List[DimensionStats],
                           summary: Dict,
                           output_file: Optional[Path]) -> None:
        """
        Output results in rich text format.
        
        Args:
            results: Filtered dimension statistics
            summary: Summary statistics
            output_file: Optional output file
        """
        output_console = Console(file=open(output_file, 'w', encoding='utf-8')) if output_file else Console()
        
        # Summary table
        summary_table = Table(title="Summary Statistics", show_header=True)
        summary_table.add_column("Metric", style="cyan", width=20)
        summary_table.add_column("Value", style="white")
        
        summary_table.add_row("Total Images", f"{summary['total_images']:,}")
        summary_table.add_row("Unique Dimensions", f"{summary['unique_dimensions']:,}")
        summary_table.add_row("Total Size", summary['total_size'])
        summary_table.add_row("Most Common", 
                             f"{summary['most_common_dimension']} ({summary['most_common_count']:,} images)")
        
        output_console.print(summary_table)
        output_console.print()
        
        # Results table
        if results:
            results_table = Table(
                title=f"Image Dimensions Analysis - {summary['total_images']:,} images",
                show_header=True
            )
            results_table.add_column("Dimensions", style="yellow", width=15)
            results_table.add_column("Count", style="green", justify="right", width=7)
            results_table.add_column("Percentage", style="blue", justify="right", width=12)
            results_table.add_column("Total Size", style="magenta", justify="right", width=12)
            results_table.add_column("Sample Files", style="white", width=40)
            
            for stats in results:
                percentage = (stats.count / summary['total_images']) * 100
                sample_files = truncate_file_list(stats.files, 3)
                
                results_table.add_row(
                    stats.dimensions_str,
                    f"{stats.count:,}",
                    f"{percentage:.1f}%",
                    stats.formatted_size,
                    sample_files
                )
            
            output_console.print(results_table)
        else:
            output_console.print("[yellow]No results match the specified criteria.[/yellow]")
        
        if output_file:
            output_console.file.close()
    
    def _output_json_format(self,
                           results: List[DimensionStats],
                           summary: Dict,
                           output_file: Optional[Path]) -> None:
        """
        Output results in JSON format.
        
        Args:
            results: Filtered dimension statistics
            summary: Summary statistics
            output_file: Optional output file
        """
        # Convert results to JSON-serializable format
        dimensions_data = []
        for stats in results:
            percentage = (stats.count / summary['total_images']) * 100 if summary['total_images'] > 0 else 0
            
            dimensions_data.append({
                "width": stats.width,
                "height": stats.height,
                "dimensions": stats.dimensions_str,
                "count": stats.count,
                "percentage": round(percentage, 1),
                "total_size_bytes": stats.total_size,
                "total_size": stats.formatted_size,
                "sample_files": stats.files[:5]  # Limit to 5 samples for JSON
            })
        
        output_data = {
            "summary": summary,
            "dimensions": dimensions_data
        }
        
        # Output JSON
        json_bytes = orjson.dumps(output_data, option=orjson.OPT_INDENT_2)
        json_str = json_bytes.decode('utf-8')
        
        if output_file:
            with open(output_file, 'w', encoding='utf-8') as f:
                f.write(json_str)
        else:
            print(json_str)
    
    def _output_yaml_format(self,
                           results: List[DimensionStats],
                           summary: Dict,
                           output_file: Optional[Path]) -> None:
        """
        Output results in YAML format.
        
        Args:
            results: Filtered dimension statistics
            summary: Summary statistics
            output_file: Optional output file
        """
        # Convert results to YAML-serializable format
        dimensions_data = []
        for stats in results:
            percentage = (stats.count / summary['total_images']) * 100 if summary['total_images'] > 0 else 0
            
            dimensions_data.append({
                "width": stats.width,
                "height": stats.height,
                "dimensions": stats.dimensions_str,
                "count": stats.count,
                "percentage": round(percentage, 1),
                "total_size_bytes": stats.total_size,
                "total_size": stats.formatted_size,
                "sample_files": stats.files[:5]  # Limit to 5 samples for YAML
            })
        
        output_data = {
            "summary": summary,
            "dimensions": dimensions_data
        }
        
        # Output YAML
        yaml_str = yaml.dump(output_data, default_flow_style=False, sort_keys=False, allow_unicode=True)
        
        if output_file:
            with open(output_file, 'w', encoding='utf-8') as f:
                f.write(yaml_str)
        else:
            print(yaml_str)
    
    def show_progress_summary(self, 
                            total_processed: int,
                            total_failed: int,
                            processing_time: float) -> None:
        """
        Show processing summary to stderr.
        
        Args:
            total_processed: Number of successfully processed files
            total_failed: Number of failed files
            processing_time: Total processing time in seconds
        """
        self.console.print(f"\n[green]Processing completed![/green]")
        self.console.print(f"  Processed: {total_processed:,} images")
        if total_failed > 0:
            self.console.print(f"  [yellow]Failed: {total_failed:,} images[/yellow]")
        self.console.print(f"  Time: {processing_time:.2f} seconds")
        if total_processed > 0:
            rate = total_processed / processing_time
            self.console.print(f"  Rate: {rate:.1f} images/second")
        self.console.print()

File: code/cli/dimensions/ruby/lib/dimensions/formatter.rb
# frozen_string_literal: true

require 'json'
require 'psych'
require 'tty-table'
require 'tty-screen'
require 'pastel'
require_relative 'utils'

module Dimensions
  # Formats and outputs dimension analysis results
  class OutputFormatter
    include Utils

    def initialize(logger)
      @logger = logger
      @pastel = Pastel.new
    end

    # Format and output results in the specified format
    def format_results(results:, format_type: 'text', sort_by: 'count',
                       min_count: 1, max_results: nil, output_file: nil)
      # Filter and sort results
      filtered_results = filter_and_sort_results(results, sort_by, min_count, max_results)

      # Generate summary statistics
      summary = generate_summary(results, filtered_results)

      # Format based on type
      case format_type
      when 'text'
        output_text_format(filtered_results, summary, output_file)
      when 'json'
        output_json_format(filtered_results, summary, output_file)
      when 'yaml'
        output_yaml_format(filtered_results, summary, output_file)
      else
        raise ArgumentError, "Unsupported format type: #{format_type}"
      end
    end

    # Show processing summary to stderr
    def show_progress_summary(total_processed:, total_failed:, processing_time:)
      warn "\n#{@pastel.green('Processing completed!')}"
      warn "  Processed: #{total_processed.to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/, '\\1,')} images"
      if total_failed.positive?
        warn "  #{@pastel.yellow("Failed: #{total_failed.to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/, '\\1,')} images")}"
      end
      warn "  Time: #{'%.2f' % processing_time} seconds"
      if total_processed.positive?
        rate = total_processed / processing_time
        warn "  Rate: #{'%.1f' % rate} images/second"
      end
      $stderr.puts
    end

    private

    # Filter and sort dimension statistics
    def filter_and_sort_results(results, sort_by, min_count, max_results)
      # Filter by minimum count
      filtered = results.values.select { |stats| stats.count >= min_count }

      # Sort based on criteria
      case sort_by
      when 'count'
        filtered.sort_by! { |x| -x.count }
      when 'dimensions'
        filtered.sort_by! { |x| [x.width, x.height] }
      else
        raise ArgumentError, "Unsupported sort criteria: #{sort_by}"
      end

      # Limit results
      filtered = filtered.first(max_results) if max_results

      filtered
    end

    # Generate summary statistics
    def generate_summary(all_results, filtered_results)
      total_images = all_results.values.sum(&:count)
      total_size = all_results.values.sum(&:total_size)
      unique_dimensions = all_results.length

      # Find most common dimension
      most_common = all_results.values.max_by(&:count)

      {
        'total_images' => total_images,
        'unique_dimensions' => unique_dimensions,
        'total_size_bytes' => total_size,
        'total_size' => format_file_size(total_size),
        'most_common_dimension' => most_common&.dimensions_str || 'None',
        'most_common_count' => most_common&.count || 0,
        'displayed_results' => filtered_results.length
      }
    end

    # Calculate percentage with zero division protection
    def calculate_percentage(count, total)
      return 0.0 if total.zero?

      (count.to_f / total) * 100
    end

    # Output results in rich text format
    def output_text_format(results, summary, output_file)
      output = if output_file
                 File.open(output_file, 'w')
               else
                 $stdout
               end

      # Summary table
      summary_table = TTY::Table.new(
        header: %w[Metric Value],
        rows: [
          ['Total Images', summary['total_images'].to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/, '\\1,')],
          ['Unique Dimensions', summary['unique_dimensions'].to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/, '\\1,')],
          ['Total Size', summary['total_size']],
          ['Most Common',
           "#{summary['most_common_dimension']} (#{summary['most_common_count'].to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/,
                                                                                          '\\1,')} images)"]
        ]
      )

      output.puts @pastel.bold('Summary Statistics')
      output.puts summary_table.render(:unicode, padding: [0, 1])
      output.puts

      # Results table
      if results.any?
        print_results(summary, results, output)
      else
        output.puts @pastel.yellow('No results match the specified criteria.')
      end

      output.close if output_file
    end

    # Print results table
    def print_results(summary, results, output)
      # Check if we can fit the full table within terminal width
      term_width = terminal_width
      use_compact_mode = should_use_compact_mode(results, term_width)

      if use_compact_mode
        print_compact_results(summary, results, output)
      else
        print_full_results(summary, results, output)
      end
    end

    # Print results in compact vertical format
    def print_compact_results(summary, results, output)
      title = "Image Dimensions Analysis - #{summary['total_images'].to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/,
                                                                               '\\1,')} images"
      output.puts @pastel.bold(title)
      output.puts 'The table size exceeds the currently set width. Using vertical orientation.'
      output.puts

      results.each_with_index do |stats, index|
        percentage = calculate_percentage(stats.count, summary['total_images'])
        sample_files = truncate_file_list(stats.files, 2)

        output.puts "#{index + 1}. #{@pastel.bold(stats.dimensions_str)}"
        output.puts "   Count: #{stats.count.to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/, '\\1,')} (#{'%.1f' % percentage}%)"
        output.puts "   Size: #{stats.formatted_size}"
        output.puts "   Sample: #{sample_files}"
        output.puts
      end
    end

    # Print results in full table format
    def print_full_results(summary, results, output)
      rows = results.map do |stats|
        percentage = calculate_percentage(stats.count, summary['total_images'])
        sample_files = truncate_file_list(stats.files, 3)

        [
          stats.dimensions_str,
          stats.count.to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/, '\\1,'),
          "#{'%.1f' % percentage}%",
          stats.formatted_size,
          sample_files
        ]
      end

      results_table = TTY::Table.new(
        header: ['Dimensions', 'Count', 'Percentage', 'Total Size', 'Sample Files'],
        rows: rows
      )

      title = "Image Dimensions Analysis - #{summary['total_images'].to_s.gsub(/(\d)(?=(\d{3})+(?!\d))/,
                                                                               '\\1,')} images"
      output.puts @pastel.bold(title)
      output.puts results_table.render(:unicode, padding: [0, 1])
    end

    # Get terminal width, fallback to 80 if unable to determine
    def terminal_width
      TTY::Screen.cols
    rescue StandardError
      80
    end

    # Determine if we should use compact mode based on estimated table width
    def should_use_compact_mode(results, terminal_width)
      return false if results.empty?

      # Estimate the width needed for the table
      # Headers: 'Dimensions', 'Count', 'Percentage', 'Total Size', 'Sample Files'
      # Add padding and borders
      max_dimensions_width = results.map { |r| r.dimensions_str.length }.max || 10
      max_count_width = results.map { |r| r.count.to_s.length }.max || 5
      max_size_width = results.map { |r| r.formatted_size.length }.max || 8
      max_sample_width = results.map { |r| truncate_file_list(r.files, 3).length }.max || 20

      # Estimate total width: columns + padding + borders
      # Using header lengths as minimums
      header_widths = [10, 5, 10, 10, 12] # Dimensions, Count, Percentage, Total Size, Sample Files
      column_widths = [
        [max_dimensions_width, header_widths[0]].max,
        [max_count_width, header_widths[1]].max,
        [10, header_widths[2]].max, # Percentage is always "XX.X%"
        [max_size_width, header_widths[3]].max,
        [max_sample_width, header_widths[4]].max
      ]

      # Add padding (2 per column) and borders (6 for unicode table)
      estimated_width = column_widths.sum + (column_widths.length * 2) + 6

      estimated_width > terminal_width
    end

    # Build dimension data structure
    def build_dimension_data(results, summary, include_all_files: false)
      results.map do |stats|
        percentage = calculate_percentage(stats.count, summary['total_images'])

        dimension_entry = {
          'width' => stats.width,
          'height' => stats.height,
          'dimensions' => stats.dimensions_str,
          'count' => stats.count,
          'percentage' => percentage.round(1),
          'total_size_bytes' => stats.total_size,
          'total_size' => stats.formatted_size
        }

        if include_all_files
          dimension_entry['files'] = stats.files
        else
          dimension_entry['sample_files'] = stats.files.first(5)
        end

        dimension_entry
      end
    end

    # Output results in JSON format
    def output_json_format(results, summary, output_file)
      dimensions_data = build_dimension_data(results, summary, include_all_files: true)
      output_data = {
        'summary' => summary,
        'dimensions' => dimensions_data
      }

      json_str = JSON.pretty_generate(output_data)

      if output_file
        File.write(output_file, json_str)
      else
        puts json_str
      end
    end

    # Output results in YAML format
    def output_yaml_format(results, summary, output_file)
      dimensions_data = build_dimension_data(results, summary, include_all_files: true)
      output_data = {
        'summary' => summary,
        'dimensions' => dimensions_data
      }

      yaml_str = Psych.dump(output_data)

      if output_file
        File.write(output_file, yaml_str)
      else
        puts yaml_str
      end
    end
  end
end
